src/gsudo/PseudoConsole/PseudoConsolePipe.cs

```diff
@@ -1,47 +1,46 @@
-﻿using Microsoft.Win32.SafeHandles;
-using System;
-using static gsudo.Native.ProcessApi;
-
-namespace gsudo.PseudoConsole
-{
-    /// <summary>
-    /// A pipe used to talk to the pseudoconsole, as described in:
-    /// https://docs.microsoft.com/en-us/windows/console/creating-a-pseudoconsole-session
-    /// </summary>
-    /// <remarks>
-    /// We'll have two instances of this class, one for input and one for output.
-    /// </remarks>
-    internal sealed class PseudoConsolePipe : IDisposable
-    {
-        public readonly SafeFileHandle ReadSide;
-        public readonly SafeFileHandle WriteSide;
-
-        public PseudoConsolePipe()
-        {
-            var sa = new SECURITY_ATTRIBUTES();
-            if (!CreatePipe(out ReadSide, out WriteSide, sa, 0))
-            {
-                throw new InvalidOperationException("failed to create pipe");
-            }
-        }
-
-        #region IDisposable
-
-        void Dispose(bool disposing)
-        {
-            if (disposing)
-            {
-                ReadSide?.Dispose();
-                WriteSide?.Dispose();
-            }
-        }
-
-        public void Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
-
-        #endregion
-    }
-}
+﻿using Microsoft.Win32.SafeHandles;
+using System;
+using static gsudo.Native.ProcessApi;
+
+namespace gsudo.PseudoConsole
+{
+    /// <summary>
+    /// A pipe used to talk to the pseudoconsole, as described in:
+    /// https://docs.microsoft.com/en-us/windows/console/creating-a-pseudoconsole-session
+    /// </summary>
+    /// <remarks>
+    /// We'll have two instances of this class, one for input and one for output.
+    /// </remarks>
+    internal sealed class PseudoConsolePipe : IDisposable
+    {
+        public readonly SafeFileHandle ReadSide;
+        public readonly SafeFileHandle WriteSide;
+
+        public PseudoConsolePipe()
+        {
+            if (!CreatePipe(out ReadSide, out WriteSide, IntPtr.Zero, 0))
+            {
+                throw new InvalidOperationException("failed to create pipe");
+            }
+        }
+
+        #region IDisposable
+
+        void Dispose(bool disposing)

```

Comment:
You don’t need this complication.
SafeFileHandle implements finalized for you.
So the only code you need is implement simple Dispose:

```
Dispose(){
  ReadSide.Dispose();
  Write side.Dispose();
}
```


The “recommended” Dispose pattern is wrong in many ways including complication and inefficiency.